src/main/java/com/tvd12/gamebox/entity/MMOGridRoom.java

```diff
@@ -0,0 +1,235 @@
+package com.tvd12.gamebox.entity;
+
+import com.tvd12.gamebox.math.Vec3;
+import lombok.Getter;
+import lombok.Setter;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class MMOGridRoom extends MMORoom {
+
+    @Getter
+    private final int maxX;
+    
+    @Getter
+    private final int maxY;
+    
+    @Getter
+    private final int maxZ;
+    
+    @Getter
+    private final int cellSize;
+    
+    private final Cell[][][] cells;
+    private final int cellRangeOfInterest;
+    private final Map<Player, Cell> cellByPlayer;
+    private final List<MMOPlayer> cellPlayerBuffer = new ArrayList<>();
+    private final Set<Cell> visitedCells = ConcurrentHashMap.newKeySet();
+
+    public MMOGridRoom(Builder builder) {
+        super(builder);
+        this.maxX = builder.maxX;
+        this.maxY = builder.maxY;
+        this.maxZ = builder.maxZ;
+        this.cellSize = builder.cellSize;
+        this.cellByPlayer = new ConcurrentHashMap<>();
+        this.cellRangeOfInterest = (int) (builder.distanceOfInterest);
+        final int maxCellX = Math.max(1, maxX / cellSize);
+        final int maxCellY = Math.max(1, maxY / cellSize);
+        final int maxCellZ = Math.max(1, maxZ / cellSize);
+        this.cells = new Cell[maxCellX][maxCellY][maxCellZ];
+    }
+
+    @Override
+    public void addPlayer(Player player) {
+        super.addPlayer(player);
+        addPlayerToCell((MMOPlayer) player);
+    }
+
+    private void addPlayerToCell(MMOPlayer player) {
+        int cellX = (int) player.getPosition().x / cellSize;
+        int cellY = (int) player.getPosition().y / cellSize;
+        int cellZ = (int) player.getPosition().z / cellSize;
+        addPlayerToCell(player, cellX, cellY, cellZ);
+    }
+    
+    private void addPlayerToCell(MMOPlayer player, int cellX, int cellY, int cellZ) {
+        Cell cell = this.cells[cellX][cellY][cellZ];
+        if (cell == null) {
+            cell = new Cell();
+        }
+        cell.setCellX(cellX);
+        cell.setCellY(cellY);
+        cell.setCellZ(cellZ);
+        cell.addPlayer(player);
+        this.cells[cellX][cellY][cellZ] = cell;
+        cellByPlayer.put(player, cell);
+    }
```

Comment:
Hàm này đang vi phạm nguyên tắc single responsibility, nghĩa là nó đang làm 2 việc trong khi tên hàm chỉ có 1 ý nghĩa là add player to cell:
1. Tạo cell
2. Add player vào cell

Cần tạo 1 hàm là createCellIfAbsent để add cell trước khi gọi addPlayerToCell